### ISSUE TITLE
The function order should be a method (of new class)

### ISSUE THREAD

**Issue by viercc** [emilypi/group-theory#15]
In the current order function quoted below, there are no groups it can return `Infinite`.
"guard against infinite cyclic cases" actually never happens, because if `gtimes k a == a` would happen,
`gtimes (k-1) a == mempty` will be encountered just before that condition holds in the `go` loop.

(Also, if I'm not mistaken it returns smallest positive integer `k` such that `a <> gtimes k a == mempty`, which is one less than the usual definition: `k` such that `gtimes k a == mempty`)

```haskell
order :: (Eq g, Group g) => g -> Order
order a = go 0 a where
  go !n g
    -- guard against ().
    | g == mempty, n > 0 = Finite n
    -- guard against infinite cyclic cases
    | g == a, n > 0 = Infinite
    | otherwise = go (succ n) (g <> a)
```

To be able to return a possibly infinite order of an element correctly, it must know what group `g` is. Making it a method allows much efficient implementation too, for example tuples.

I'd suggest something like this:

```haskell
class (Group g, Eq g) => DiscreteGroup g where
    order :: g -> Order
    order = {- the above default impl. -}
```

It implies `Eq g` because `order (x <> invert y) == Finite 1` is equivalent to `x == y`.

**Comment by emilypi** [comment]
`order` got pretty crufty pretty quickly for sure. Originally, the second guard was there to guard against invalid instances that were removed in later PR's, since it's never the case that it's tripped, as you correctly state. Ideally, `order` should be a function on `Cyclic` groups along with `generator` imo, since you always have the iso `C_n -> Z/nZ` or `C -> Z` when the group is infinite.

**Comment by viercc** [comment]
> function on `Cyclic` groups along with `generator`

I'm not sure, because many groups are not cyclic, notably `Sum Rational` isn't. Even a direct product of two cyclic groups is not cyclic in general.

But `order` function for them are simple enough:

```haskell
-- I'll use the hypothetical class DiscreteGroup.
-- I don't know if the name is a good fit though.
instance DiscreteGroup (Sum Rational) where
    order 0 = Finite 1
    order _ = Infinite
instance (DiscreteGroup a, DiscreteGroup b) => DiscreteGroup (a,b) where
    order (a,b) = order a `lcmOrder` order b

lcmOrder :: Order -> Order -> Order
lcmOrder (Finite m) (Finite n) = Finite (lcm m n)
lcmOrder _ _ = Infinite
```

To be honest, my experience on using groups in Haskell is limited.
Only for doing some calculation in Haskell with "cute" abstraction, and groups were not of absolute necessity for me.
I can't judge how `order` function is useful enough to justify making a type class just for it.

**Comment by emilypi** [comment]
Ah I see what you were talking about - `order` as separate in the sense that you think a class dedicated to defining order for some *individual elements*. Okay. That makes sense to me, and it makes sense that we'd want this to be ad-hoc so we can label finite and infinite parts. Yeah that sounds okay to me.

**Comment by emilypi** [comment]
merged in #20
